UniversalAlgebra/Homomorphisms/Setoid.lagda

```diff
@@ -0,0 +1,307 @@
+  B = 𝕌[ 𝑩 ]
+
+ is-monomorphism : (A → B) → Type _
+ is-monomorphism g = is-homomorphism 𝑨 𝑩 g × IsInjective g
+
+ mon : Type(𝓞 ⊔ 𝓥 ⊔ α ⊔ β)
+ mon = Σ[ g ∈ (A → B) ] is-monomorphism g
+
+ is-epimorphism : (A → B) → Type _
+ is-epimorphism g = is-homomorphism 𝑨 𝑩 g × IsSurjective g
+
+ epi : Type _
+ epi = Σ[ g ∈ (A → B) ] is-epimorphism g
+
+
+-- The "hom reduct" of a mon
+-- N.B. 𝑨 explicit, 𝑩 implicit
+module _ {α ρᵃ : Level} (𝑨 : SetoidAlgebra α ρᵃ)
+         {β ρᵇ : Level} {𝑩 : SetoidAlgebra β ρᵇ} where
+
+ mon-to-hom : mon 𝑨 𝑩 → hom 𝑨 𝑩
+ mon-to-hom ϕ = ∣ ϕ ∣ , fst ∥ ϕ ∥
+
+
+-- The "hom reduct" of an epi
+-- N.B. 𝑨 implicit, 𝑩 explicit
+module _ {α ρᵃ : Level} {𝑨 : SetoidAlgebra α ρᵃ}
+         {β ρᵇ : Level} (𝑩 : SetoidAlgebra β ρᵇ) where
+
+ epi-to-hom : epi 𝑨 𝑩 → hom 𝑨 𝑩
+ epi-to-hom ϕ = ∣ ϕ ∣ , fst ∥ ϕ ∥
+
+\end{code}
+
+
+
+#### Kernels of homomorphisms for SetoidAlgebras
+
+\begin{code}
+
+
+module _ {α ρᵃ : Level} (𝑨 : SetoidAlgebra α ρᵃ)
+         {β ρᵇ : Level} (𝑩 : SetoidAlgebra β ρᵇ)
+         where
+ private
+  A = 𝕌[ 𝑨 ]
+  B = 𝕌[ 𝑩 ]
+
+
+ homker-comp : swelldef 𝓥 β → (h : hom 𝑨 𝑩) → 𝑨 ∣≈ (ker ∣ h ∣)
+ homker-comp wd h f {u}{v} kuv = ∣ h ∣((f ̂ 𝑨) u)   ≡⟨ ∥ h ∥ f u ⟩
+                                     (f ̂ 𝑩)(∣ h ∣ ∘ u) ≡⟨ wd(f ̂ 𝑩)(∣ h ∣ ∘ u)(∣ h ∣ ∘ v)kuv ⟩
+                                     (f ̂ 𝑩)(∣ h ∣ ∘ v) ≡⟨ (∥ h ∥ f v)⁻¹ ⟩
+                                     ∣ h ∣((f ̂ 𝑨) v)   ∎
+
+-- NOT WORKING YET
+--  kercon : swelldef 𝓥 β → hom 𝑨 𝑩 → Con 𝑨 -- {α}{β}
+--  kercon wd h = ker ∣ h ∣ , {!!} -- mkcon (ker-IsEquivalence ∣ h ∣)(homker-comp wd {𝑩} h)
+
+--  kerquo : swelldef 𝓥 β → hom 𝑨 𝑩 → SetoidAlgebra _ _ -- (α ⊔ lsuc β) _
+--  kerquo wd h = 𝑨 ╱ (kercon wd h)
+
+
+-- ker[_⇒_]_↾_ : {α ρᵃ : Level} (𝑨 : SetoidAlgebra α ρᵃ)
+--               {β ρᵇ : Level} (𝑩 : SetoidAlgebra β ρᵇ)
+--  →            hom 𝑨 𝑩 → swelldef 𝓥 β → SetoidAlgebra _ _ --  (α ⊔ lsuc β) _
+-- ker[ 𝑨 ⇒ 𝑩 ] h ↾ wd = kerquo 𝑨 𝑩 wd h
+
+\end{code}
+
+
+
+
+
+
+#### Homomorphism Decomposition for SetoidAlgebras
+
+
+If `τ : hom 𝑨 𝑩`, `ν : hom 𝑨 𝑪`, `ν` is surjective, and `ker ν ⊆ ker τ`, then there exists `φ : hom 𝑪 𝑩` such that `τ = φ ∘ ν` so the following diagram commutes:
+
+```
+𝑨 --- ν ->> 𝑪
+ \         .
+  \       .
+   τ     φ
+    \   .
+     \ .
+      V
+      𝑩
+```
+
+\begin{code}
+
+module _ {α ρᵃ : Level} {𝑨 : SetoidAlgebra α ρᵃ}
+         {β ρᵇ : Level} (𝑩 : SetoidAlgebra β ρᵇ)
+         {γ ρᶜ : Level} {𝑪 : SetoidAlgebra γ ρᶜ} where
+
+ private
+  A = 𝕌[ 𝑨 ]
+  B = 𝕌[ 𝑩 ]
+  C = 𝕌[ 𝑪 ]
+
+ open import Axiom.Extensionality.Propositional    using    ()
+                                                  renaming (Extensionality to funext)
+
+ HomFactor : funext α β → swelldef 𝓥 γ
+  →          (τ : hom 𝑨 𝑩)(ν : hom 𝑨 𝑪)
+  →          kernel ∣ ν ∣ ⊆ kernel ∣ τ ∣ → IsSurjective ∣ ν ∣
+             --------------------------------------------------
+  →          Σ[ φ ∈ (hom 𝑪 𝑩)] ∣ τ ∣ ≡ ∣ φ ∣ ∘ ∣ ν ∣
```

Comment:
This is definitely one definition that should be changed to use pointwise equality. "\=o" in emacs and in stdlib. That's the whole point with having 𝑩 be a Setoid!

Reply:
Okay, I'll change this now.